### PR TITLE
Fix pressure dependent network generation in RMG jobs

### PR DESCRIPTION
**Pull request by mliu49**
### Note
I consider this to be a major bug and therefore plan on making a patch release (2.4.1) once this is merged. My initial thought is to only include this fix, but I'm also considering whether to simply include everything that's been merged since the 2.4 release. Thoughts welcome.

### Motivation or Problem
An apparent slowdown in RMG model generation was observed recently. The following plots were from a dodecane pyrolysis model. The exact commits are not too important here. This was just the initial observation that something major was different.

Screenshot from 2019-07-16 16-40-31

Further investigation showed that the problem was related to pdep networks. In the "after" model, RMG would continuously explore pdep networks on every iteration instead of adding new core species. Network leak rate ratios were all absurd values around 1e100~1e300.

After a ton of debugging, the core problem was identified as incorrect identification of `newSpecies` when calling `processNewReactions`. The `newSpecies` argument was used by `addReactionToUnimolecularNetworks` to identify the direction in which to process a new reaction, and was intended to correspond to the core species which was used to generate the reaction.

In #1459, the selection of `newSpecies` was changed following changes to no longer deflate and inflate reactions. The new method was to simply find the first core species in the reaction. However, this did not always correspond to the species used to generate the reaction. The result was improper creation of pdep networks.

### Description of Changes
The goal was to fix how we identified `newSpecies`. The chosen implementation was to return the species tuples used to generate the reactions from `react_all` since they directly map the generated reactions to the core species used to generate them. Note that we make copies of every species before reaction generation, so the actual species contained in the reactions do not point to the actual core species.

Changes:
- `react` generates a list of lists due to `map`, return that as is instead of chaining the lists together
- in `react_all`, return the list of lists from react as is, along with the list of species tuples provided to `react`
- in enlarge, match reactions with their species tuples, and take the species in the tuple with the _lowest_ index as `newSpecies` (chosen for consistency, I don't think it really matters, since it seems that `addReactionToUnimolecularNetworks` only uses it decide on the direction to process the reaction)
- add a `flatten` option to `react` primarily to simplify cases where you just want a flat list of reactions
- update related unit tests to handle the changes in return values
- organize imports in rmgpy.rmg.model

### Testing
I used the dodecane model mentioned above to debug and test. I'm currently running it for longer to confirm.

### Reviewer Tips
Read my explanation above, read the code, ask questions, maybe run a test case?

### CODE REVIEW

**Comment by mliu49** [comment]
Made the following changes following offline discussion:
- Removed the `flatten` argument since it's only used for unit tests. Instead, flatten in the unit tests.
- Do not sort the species tuples since `newSpecies` is only used to determine reaction direction.
- Check that the reaction list is non-empty before processing
- More thoroughly update docstrings

I went ahead and rebased on master.

**Comment by rwest** [comment]
Well done debugging that. 
Should we add a test that would have caught this regression?

**Comment by mliu49** [comment]
I created a new functional test for `enlarge` with pressure dependence. I tried to make it as minimal as possible, but it was quite difficult.

I did confirm that it fails on master.

**Comment by alongd** [comment]
The changes look OK to me

**Comment by mliu49** [comment]
@amarkpayne, @ajocher, do you have any final comments?

**Comment by ajocher** [comment]
Looks good to me.

File: rmgpy/rmg/modelTest.py [hunk 1]
```diff
@@ -579,5 +582,123 @@ def tearDownClass(cls):
...
+    @classmethod
+    def setUpClass(cls):
+        """
+        A method that is run ONCE before all unit tests in this class.
+        """
+        cls.dirname = os.path.abspath(os.path.join(os.path.dirname(__file__), 'temp'))
+        print cls.dirname
+        os.makedirs(os.path.join(cls.dirname, 'pdep'))
+
+        TESTFAMILY = 'R_Recombination'
```
[the remark this answers is not in the document]
Comment (mliu49) [reply rmgpy/rmg/modelTest.py]: Good point. I copied the method from another test class and didn't change the variable name. It should be all lowercase.

I also just noticed that I left a print statement in...

[the remark this answers is not in the document]
Comment (mliu49) [reply rmgpy/rmg/modelTest.py]: I put a commit fixing these into #1659.

**Pull request merged**